Add plot.violin()

JS3xton opened the pull request:
Add `plot.violin()`, which easily and flexibly illustrates violin plots from flow cytometry data.

Closes https://github.com/taborlab/FlowCal/issues/332.

Still TODO: create tutorial images and update image links in `/doc/python_tutorial/plot.rst`.

<br>

#### Wishlist:
- Support logicle scale. (y-axis only?)
- Add `density` parameter? (like `normed_area` parameter of `plot.hist1d`)
- Support `pd.Series` and `pd.Dataframe` inputs? (probably better left to https://github.com/taborlab/FlowCal/issues/76)

<br>

#### Testing:
In both `Python 3.8 + Anaconda 2020.02` and `Python 2.7 + Anaconda 4.4.0`:
- All unit tests pass.
- The updated example analysis scripts run without error (`analyze_no_mef.py`, `analyze_mef.py`, and `analyze_excel_ui.py` in `/examples`).
- The plot violin tutorial commands run without error (in `/doc/python_tutorial/plot.rst`).

<br>

#### Thoughts on some failure modes:

Supporting the `channel` parameter, which all `plot` functions offer, [was a bit tricky](https://github.com/JS3xton/FlowCal/blob/90fed69d6e3ff3e0eb954041473820f5a3ef84e3/FlowCal/plot.py#L1704) given how flexible my original code was to different inputs. If `plot.violin()` is used correctly, it works [as described](https://github.com/JS3xton/FlowCal/blob/90fed69d6e3ff3e0eb954041473820f5a3ef84e3/FlowCal/plot.py#L1507). There are some potentially common failure modes, though:
- `plot.violin(fcs_data)` (without `channel`) will attempt to plot each row of `fcs_data` as a violin. This happens because `plot.violin()` understands `fcs_data` to be a sequence of 1D sequences (which it technically is). This can be bad for `fcs_data` with many rows.
- `plot.violin([fcs_data1,fcs_data2,fcs_data3])` (again without `channel`) will concatenate all channels for each `fcs_data` and attempt to plot them, which usually fails trying to calculate `log(0)`. This happens because the input is understood to be a sequence of 1D arrays, and each 1D array is [iterated over using the `.flat` numpy iterator feature](https://github.com/JS3xton/FlowCal/blob/90fed69d6e3ff3e0eb954041473820f5a3ef84e3/FlowCal/plot.py#L1941).

I chose not to address these failure modes, but I still wanted to point them out.

JS3xton committed:
Copied over plot_violin() code.

JS3xton committed:
Modified plot.violin to conform to other plot functions.

JS3xton committed:
Fix old module reference.

matplotlib was previously imported as "mpl" in plot_violin code. In
FlowCal, matplotlib is not imported under this alias. This commit
corrects "mpl" references to "matplotlib".

JS3xton committed:
Minor plot.violin bug and documentation fixes.

-Using a pd.Series for `positions` with a 0 position would fail
 because using `0 in positions` would check if 0 was in the index,
 not the positions.
-x ticks would not be properly labeled under some circumstances
 because the plot was not properly drawn initially.

JS3xton committed:
Added violin plots to example analysis scripts.

JS3xton committed:
Fix plot.violin bug where channel not properly extracted.

Fix plot.violin bug where channel was not extracted from zero, min,
or max data.

JS3xton committed:
Improve robustness of plot.violin drawing user-specified model.

JS3xton committed:
Improved plot.violin x-tick labeling.

-Made custom tick locators and formatters aware of min, max, and
 zero violins.
-Exposed min, max, and zero violin tick labels for user to specify
 if so desired.

JS3xton committed:
Add violin plot section to plotting tutorial.

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        try:
+            data = [d[:,channel] for d in data]
+        except TypeError:
+            data = [[row[channel] for row in d] for d in data]
```
What situation is this supposed to cover? If every `data[i]` is a list of lists and `channel` is an int?

JS3xton replied:
Yeah, pretty much.

In the words of the docstring, it permits `data` to be a sequence of ND sequences where the ND sequences cannot be indexed via multi-dimensional indexing (i.e. `ndseq[:,channel]`). ND sequences comprised of simpler data structures, like lists and tuples, are indeed prototypical examples.

Another data structure I had in mind was a `DataFrame`, which will *not* work as currently written. Iterating over a `DataFrame` iterates over its column names, not its rows. I figured supporting a `DataFrame` should be relegated to https://github.com/taborlab/FlowCal/issues/76, though. And I suspect a similar failure would arise with ND sequences as dictionaries mapping from channel to sequence.

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        positions_length = len(positions)
+
+        if logx_zero_data is None:
+            logx_zero_data = zero_data
```
Maybe there needs to be a warning here if `log_zero_data` is not None? In this case, `data[zero_idx]` will be ignored, which may not be immediately obvious to the user.

JS3xton replied:
Yeah, I think that's a good idea if we keep this behavior. The silent preference for `log_zero_data` over `data[zero_idx]` could otherwise be dubious.

You originally weren't sure about this automatic separation of the zero violin, though. I offered the alternative of ignoring `data[zero_idx]` if the x-axis is logarithmic, but we never resolved the issue.

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        Number of bins to bin population members into along the y-axis.
+        Ignored if `y_bin_edges` is specified.
+    y_bin_edges : sequence of scalars, sequence of sequence of scalars, or
+                  mapping to sequence of scalars, optional
```
Do we expect any mapping other than dictionaries? Dictionaries appear to be the only standard mapping in python. I would suggest simplifying the terminology to "dictionary".

castillohair replied:
Same with "sequence" -> "list", unless we're expecting an array, for which there is the "array_like" term. This would make it more consistent with the rest of `Flowcal.plot`.

JS3xton replied:
I don't know what you mean by "expect". If robust, the function should work with any reasonable input. My nomenclature is intended to be as general as possible and is inspired by Python's [Abstract Base Classes (ABCs)](https://docs.python.org/3/library/collections.abc.html), which include `Sequence` and `Mapping`. Moreover, the `Sequence` and `Mapping` ABCs are used pretty extensively throughout the code. So I guess you could say the code expects anything that is a `Sequence` or a `Mapping`.

More concretely, `Sequence` includes `list` _and_ `tuple`, and `Mapping` includes `dict` and `OrderedDict`.

It appears `ndarray` doesn't quite make the cut for `Sequence`: https://github.com/numpy/numpy/issues/2776, https://github.com/numpy/numpy/issues/7315. This could actually be an issue in a few places (e.g. specifying `upper_trim_fraction` or `lower_trim_fraction` as an `ndarray` of `floats`); I didn't test `ndarrays` extensively.

And `array_like` can actually be quite vague: https://stackoverflow.com/a/40380014.

Ultimately, I wanted `tuple` to work as well as `list` and `ndarray` anywhere I said "sequence". `pd.Series` and `pd.DataFrame` would also be nice, but I consider those the purview of https://github.com/taborlab/FlowCal/issues/76.

castillohair replied:
I understand that you're trying to be general with these abstract classes, but these `Sequence` and `Mapping` names are not widespread and  make the documentation more confusing than it needs to be. I didn't even know of their existence, and only knew where to look them up after seeing that they come from `collections` after looking at the code. Furthermore, these are not used anywhere else in `FlowCal`, and it is very inconsistent to have one single function that suddenly uses this nomenclature. I'd rather use `list` and `dictionary` and drop the theoretical compatibility with other `Sequence` and `Mapping` objects, and gain in terms of documentation readability.

JS3xton replied:
Are you advocating for dumbing down the code? Or just the documentation?
<br>
> Furthermore, these are not used anywhere else in FlowCal, and it is very inconsistent to have one single function that suddenly uses this nomenclature.

I agree here. Although I'd prefer improving the robustness everywhere else. There are several places where `isinstance(..., list)` is used, which unnecessarily prevents other sequence data structures (namely `tuple`). The documentation usually specifies `list`, at least, but it still feels unnecessarily restrictive. This would obviously have to be a separate issue, though.

We also appear to use `numpy array` and `array_like` interchangeably in some spots? Which is kinda confusing.

castillohair replied:
> Are you advocating for dumbing down the code? Or just the documentation?

Just for the documentation and error messages. The code looks fine and is probably more robust as you said. We can claim we provide support for the basic data structures and silently provide support for the more general abstract classes.

JS3xton replied:
Ahh, OK cool. Yeah, I can get behind that.

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+            except TypeError:
+                msg  = "`y_bin_edges` should be iterable sequence or sequence"
+                msg += " of sequences"
+                raise TypeError(msg)
```
Sounds like this should have been checked outside the `if xscale == 'log' and 0 in list(positions):` first.

JS3xton replied:
`y_bin_edges` is understood on a per-violin basis, and the behavior is slightly different for `data` violins versus `min`, `max`, and `logx_zero` violins. So there is no global check of `y_bin_edges` like with some of the other parameters above. The only reason it's checked here is to cover the edge case where the user specified a sequence of individual `y_bin_edges` for each violin and we need to extract the `y_bin_edges` for the zero violin (the surrounding code does something similar for the other parameters).

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        violin's position, 'data'. Min, max, and zero kwargs can be specified
+        via the 'min', 'max', and 'logx_zero' keys, respectively. If
+        `violin_kwargs` is a sequence of mappings, min, max, and zero violins
+        cannot be used. Default = {'facecolor':'gray', 'edgecolor':'black'}.
```
This needs examples. It's a little hard to understand.

JS3xton replied:
Agreed.

Several parameters behave similarly in this regard. I'm not sure if it's better to make separate examples for all of them or to reorganize the docstring and consolidate like parameters. E.g. adding a "Per-Violin Parameters" section and explaining the different ways they can be specified.

castillohair replied:
The problem with the  "Per-Violin Parameters" idea is that I don't know how `sphinx` would parse it when generating the documentation. I'm OK with having examples for every parameter, even if that means that there is some repetition.

JS3xton replied:
`sphinx` parses "Other Parameters" sections for many functions in [the `plot` module](https://flowcal.readthedocs.io/en/latest/reference/FlowCal.plot.html). Is there any reason to believe a "Per-Violin Parameters" section (or an "Advanced Parameters" section for `min_data`, `max_data`, and `log_zero_data`, as discussed previously) would be treated any differently?

castillohair replied:
There is a documented "Other Parameters" section in the [`numpydoc` specification](https://numpydoc.readthedocs.io/en/latest/format.html#sections).

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        and zero `draw_summary_stat_kwargs` can be specified via the 'min',
+        'max', and 'logx_zero' keys, respectively. If
+        `draw_summary_stat_kwargs` is a sequence of mappings, min, max, and
+        zero violins cannot be used. Default = {'color':'black'}.
```
This also needs examples.

JS3xton replied:
see response above

castillohair commented on FlowCal/plot.py:
```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+                     draw_summary_stat_fxn=draw_summary_stat_fxn,
+                     draw_summary_stat_kwargs=v_draw_summary_stat_kwargs)
+
+        if draw_model:
```
Why is this inside the `for idx in range(data_length):`? Does this mean that the model plot is made every time a new violin is plotted?

JS3xton replied:
Hmm, good catch. Yeah, that should probably be unindented once.

JS3xton committed:
Merge branch 'develop' into plot-violin

JS3xton committed:
Fix small bug caused by merging with develop.

JS3xton committed:
Fixed minor indentation error.

De-indented code that plots model for violin plot, which previously
erroneously plotted model multiple times.

JS3xton committed:
Rename _plot_violin() to _plot_single_violin().

JS3xton committed:
Split `plot.violin()` into two functions.

Made simpler `plot.violin()`.
-removed mathematical model
-removed min and max violins
-automatically extracted position 0 w/ log x-axis (removed
 `logx_zero_data` parameter)
-renamed `data_xlim` to `xlim`
-simplified `violin_kwargs` (and other param) specifications
 (removed mapping of mappings option)
-simplified documentation ("mapping" -> "dict", "sequence" ->
 "list" or "array")

Made more complex `plot.violin_dose_response`.
-automatically extracts position 0 w/ log x-axis (no
 `logx_zero_data` parameter)
-simple `violin_kwargs` (and other param) specifications (no
 mapping of mappings option)
-added min and max versions of the following params: violin_kwargs
 draw_summary_stat_kwargs, y_bin_edges, upper_trim_fraction, and
 lower_trim_fraction
-simplified documentation ("mapping" -> "dict", "sequence" ->
 "list" or "array")
-renamed `data_xlim` to `xlim`

JS3xton committed:
Add `density` parameter to violin plot functions.

JS3xton committed:
Add logicle y-scale option to violin plot functions.

JS3xton committed:
Changed example dataset and scripts.

JS3xton committed:
Minor update to plot tutorial.

Updated plot tutorial to reference both `plot.violin` and
`plot.violin_dose_response`. The tutorial still needs to be updated
to use the new DAPG data set, though, instead of the old IPTG data
set.

castillohair commented on FlowCal/plot.py:
```diff
@@ -1047,6 +1288,1460 @@ def hist1d(data_list,
+                                                    'right_side_x',
+                                                    'y'))
+
+def _plot_single_violin(violin_position,
```
Could you write a short docstring for this function? This is not a user-facing function, but we should have something for people looking to modify this function in the future.

JS3xton replied:
Done in 5336718ae4f55574909594ccbea607483e9e7e8f.

castillohair commented on examples/analyze_excel_ui.py:
```diff
@@ -85,50 +85,153 @@
+    cmap = mpl.cm.get_cmap('cividis')
+    norm = mpl.colors.LogNorm(vmin=1e0, vmax=500)
+    colors = [cmap(norm(dapg_i)) if dapg_i > 0 else cmap(0.0)
+              for dapg_i in dapg]
```
I would rather remove this colormap section. I don't think it serves to illustrate anything unique about FlowCal, and adds lines to example files which are now several hundred lines long. I don't feel too strongly about it though.

JS3xton replied:
Yeah, I agree the custom colormap lengthens the example. I did this because the default colormap renders a barely perceptible yellow histogram, which I think is bad data vizualization technique:

| `develop` (3affd34e00060b9212cfc9093aa9695763aa8e96) | `master` (5e187b0975499cc7887a537df81c511b0ee4798a) |
|:---:|:---:|
| <img src="https://user-images.githubusercontent.com/2179825/94490003-2f53f980-01ab-11eb-8b7a-d690070ec0e7.png" width="200" height="150"> | <img src="https://user-images.githubusercontent.com/2179825/94490116-6cb88700-01ab-11eb-849f-3314074d9088.png" width="200" height="150"> |

If we don't use a custom colormap, the current example becomes:

| `plot-violin` (5336718ae4f55574909594ccbea607483e9e7e8f) | `plot-violin` (no colormap) |
|:---:|:---:|
| <img src="https://user-images.githubusercontent.com/2179825/94490446-e2bcee00-01ab-11eb-8af6-74992f330722.png" width="200" height="150"> | <img src="https://user-images.githubusercontent.com/2179825/94490464-eea8b000-01ab-11eb-9c63-ef30fd7a6cbf.png" width="200" height="150"> |

If you think the simpler example is more important, though, I can remove the custom colormaps.

castillohair replied:
Again, I don't feel too strongly about it. But the histograms with the default colormap look fine although not great when you open the figure. Another argument is that fixing a crappy default colormap should not be this PR's problem, so we should just use the default and not worry about it. But this is an example and which colormap we use doesn't really matter.

I don't know. I'm fine if you leave it. Seems to be the alternative that requires the least effort from where we are.

JS3xton replied:
Sounds good; I'll leave it.

castillohair commented on examples/analyze_excel_ui.py:
```diff
@@ -85,50 +85,153 @@
+    plt.xlim((0,5e4))
     plt.xlabel('FL1  (Molecules of Equivalent Fluorescein, MEFL)')
-    plt.legend(['{} $\mu M$ IPTG'.format(i) for i in iptg],
+    plt.legend(['{} $\mu M$ DAPG'.format(i) for i in dapg],
```
Too many decimal places. Replace `{}` with `{:.1f}` or `{:.2f}`.

JS3xton replied:
Done in ab40d38358da8af0b33bba3c532ca8d5b4b173df.

JS3xton committed:
Added docstring for `plot._plot_single_violin()`.

JS3xton committed:
Reduced decimal places in legend of example plot.

JS3xton committed:
Changed example plot summary statistic to mean.

JS3xton committed:
Allow horizontal violin plots with `plot.violin()`.

Add `vert` parameter to `plot.violin()`, which allows violins to be
plotted horizontally. Also generalized some parameter names for
both `plot.violin()` and `plot.violin_dose_response()` (i.e.,
removed references to axes, e.g., `y_bin_edges` -> `bin_edges`,
etc.).

castillohair commented on FlowCal/plot.py:
```diff
@@ -1047,6 +1288,1578 @@ def hist1d(data_list,
+           yscale=None,
+           xlim=None,
+           ylim=None,
+           vert=True,
```
There is no docstring for this new parameter.

JS3xton replied:
Ooo, good catch. Fixed in 0de9b02d086a5d9dd939a73af1050ca786a34a7d.

JS3xton committed:
Add `vert` to `plot.violin()` docstring.

castillohair committed:
Updated documentation images and API tutorials.

JS3xton committed:
Minor update to tutorial violin plots.

JS3xton committed:
Updated tutorial images for the Excel UI.

castillohair committed:
Updated figures in the 'fundamentals' section of the documentation.

JS3xton committed:
Updated tutorial image for "Getting Started" section.

castillohair committed:
Added version policy to documentation.
